src/couch_index/test/couch_index_ddoc_updated_tests.erl

```diff
@@ -0,0 +1,134 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_index_ddoc_updated_tests).
+
+-include_lib("couch/include/couch_eunit.hrl").
+-include_lib("couch/include/couch_db.hrl").
+
+setup() ->
+    DbName = ?tempdb(),
+    fake_mem3(DbName),
+    fake_index(),
+
+    DbNames = [mem3:name(Sh) || Sh <- mem3:local_shards(mem3:dbname(DbName))],
+    [Db1| RestDbs] = lists:map(fun(DBName) ->
+        {ok, DB} = couch_db:create(DBName, [?ADMIN_CTX]),
+        DB
+    end, DbNames),
+
+    % create a DDoc on the 1st Db
+    DDocID = <<"idx_name">>,
+    DDocJson = couch_doc:from_json_obj({[
+        {<<"_id">>, DDocID},
+        {<<"value">>, 1}
+    ]}),
+    {ok, _Rev} = couch_db:update_doc(Db1, DDocJson, []),
+    {ok, Db} = couch_db:reopen(Db1),
+    {ok, DDoc} = couch_db:open_doc(Db, DDocID, [ejson_body, ?ADMIN_CTX]),
+    Dbs = [Db | RestDbs],
+    {Dbs, DDoc}.
+
+
+teardown({Dbs, _DDoc}) ->
+    lists:foreach(fun(Db) ->
+        couch_db:close(Db),
+        couch_server:delete(Db#db.name, [?ADMIN_CTX])
+    end, Dbs),
+    (catch meck:unload(mem3)),
+    (catch meck:unload(test_index)),
+    ok.
+
+
+ddoc_update_test_() ->
+    {
+        "Check ddoc update actions",
+        {
+            setup,
+            fun() -> test_util:start_couch([]) end, fun test_util:stop_couch/1,
+            {
+                foreach,
+                fun setup/0, fun teardown/1,
+                [
+                    fun ddoc_update/1
+                ]
+            }
+        }
+    }.
+
+
+ddoc_update({[Db | _] = Dbs, DDoc}) ->
+    ?_test(begin
+        N = length(Dbs),
+        DDocID = DDoc#doc.id,
+
+        % run couch_index process for each Db
+        lists:foreach(fun(DB) ->
+            couch_index_server:get_index(test_index, DB, DDoc)
+        end, Dbs),
+        IndexesBefore = ets:match_object(
+            couchdb_indexes_by_db, {'$1', {DDocID, '$2'}}),
+        ?assertEqual(N, length(IndexesBefore)),
+
+        % update ddoc
+        DDocJson2 = couch_doc:from_json_obj({[
+            {<<"_id">>, DDocID},
+            {<<"value">>, 2},
+            {<<"_rev">>, couch_doc:rev_to_str(DDoc#doc.revs)}
+        ]}),
+        {ok, _} = couch_db:update_doc(Db, DDocJson2, []),
+
+        % assert that all index processes exited after ddoc updated
+        couch_index_server:handle_db_event(
+            Db#db.name, {ddoc_updated, DDocID}, {st, ""}),
+        timer:sleep(1000),
```

Comment:
Here is what works locally:
```
--- a/src/couch_index/test/couch_index_ddoc_updated_tests.erl
+++ b/src/couch_index/test/couch_index_ddoc_updated_tests.erl
@@ -72,13 +72,17 @@ ddoc_update({[Db | _] = Dbs, DDoc}) ->
         DDocID = DDoc#doc.id,

         % run couch_index process for each Db
+        ok = meck:reset(test_index),
         lists:foreach(fun(DB) ->
             couch_index_server:get_index(test_index, DB, DDoc)
         end, Dbs),
-        IndexesBefore = ets:match_object(
-            couchdb_indexes_by_db, {'$1', {DDocID, '$2'}}),
+        IndexesBefore = get_indexes_by_ddoc(DDocID),
+
         ?assertEqual(N, length(IndexesBefore)),

+        AliveBefore = lists:filter(fun erlang:is_process_alive/1, IndexesBefore),
+        ?assertEqual(N, length(AliveBefore)),
+
         % update ddoc
         DDocJson2 = couch_doc:from_json_obj({[
             {<<"_id">>, DDocID},
@@ -88,15 +92,30 @@ ddoc_update({[Db | _] = Dbs, DDoc}) ->
         {ok, _} = couch_db:update_doc(Db, DDocJson2, []),

         % assert that all index processes exited after ddoc updated
+        ok = meck:reset(test_index),
         couch_index_server:handle_db_event(
             Db#db.name, {ddoc_updated, DDocID}, {st, ""}),
-        timer:sleep(1000),
-        IndexesAfter = ets:match_object(
-            couchdb_indexes_by_db, {'$1', {DDocID, '$2'}}),
+        ok = meck:wait(N, test_index, init, ['_', '_'], 5000),
+        IndexesAfter = get_indexes_by_ddoc(DDocID),
         ?assertEqual(0, length(IndexesAfter)),
+
+        %% make sure previously running indexes are gone
+        AliveAfter = lists:filter(fun erlang:is_process_alive/1, IndexesBefore),
+        ?assertEqual(0, length(AliveAfter)),
+
         ok
     end).

+get_indexes_by_ddoc(DDocID) ->
+    Indexes = ets:match_object(
+        couchdb_indexes_by_db, {'$1', {DDocID, '$2'}}),
+    [get_index(DbName, Sig) || {DbName, {_DDocID, Sig}} <- Indexes].
+
+get_index(DbName, Sig) ->
+    case ets:lookup(couchdb_indexes_by_sig, {DbName, Sig}) of
+        [{_, Pid}] -> Pid;
+        _ -> false
+    end.
 ```
Although I would move get_indexes_by_ddoc and get_index into couch_index_server module.